content/reference/v1beta1/repeatable.md

```diff
@@ -0,0 +1,278 @@
+
+`yamlPath` **must** end with an array.
+```yaml
+    template:
+    - apiVersion: v1
+      kind: Service
+      name: my-service
+      namespace: my-app
+      yamlPath: 'spec.ports[0]'
+```
+
+If the `yamlPath` field is blank, the entire YAML document matching the `template` will be replaced with a copy for each of the repeatable item entries.  The `metadata.name` field of the new doc will reflect the repeatable item `key`.
+
+## Templating
+
+The repeat items are called with the delimeters `repl[[ .itemName ]]` or `[[repl .itemName ]]`.  These delimiters can be placed anywhere inside of the `yamlPath` target node.
+```yaml
+    - port: '{{repl ConfigOption "[[repl .service_ports ]]" | ParseInt }}'
+```
+This repeatable templating is not compatible with sprig templating functions.  It is designed for inserting repeatable `keys` into the manifest. Repeatable templating **can** be placed inside of Replicated config templating.
+
+## Ordering
+
+Repeatable templates are processed before config template rendering.
+
+Repeatable items are processed in order of the template targets in the Config Spec file.  Effectively, this ordering is from the top of the Config Spec, by Config Group, by Config Item, and then by template target.
+
+```yaml
+    - name: ports
+      items:
+      - name: service_ports
+        title: Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1 #processed first
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath: 'spec.ports[0]'
+        - apiVersion: v1 #processed second
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath:
+    {...}
+      - name: other_ports
+        title: Other Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1 #processed third
+          kind: Service
+          name: my-other-service
+          namespace: my-app
+    {...}
+    - name: deployments
+      items:
+      - name: deployment-name
+        title: Deployment Names
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: apps/v1 #processed fourth
+          kind: Deployment
+          name: my-deployment
+          namespace: my-app
+          yamlPath:
+    {...}
+```
+
+## Repeatable Item Example for a YamlPath
+**Config spec**
+```yaml
+    - name: ports
+      items:
+      - name: service_ports
+        title: Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath: 'spec.ports[0]'
+        valuesByGroup:
+          ports:
+            port-default-1: "80"
+            port-default-2: "443"
```

Comment:
Just making sure I understand the basics here… In this example, when the end user is installing the app and has to enter the configuration values, would this present them with two port values by default (80 and 443) and they could click a plus button to add additional ones?

Where does the key name come from? Like in this case you have `port-default-1, and `port-default-2`, but if the end user added a third, what would it be called? I assume it would be generated programmatically.

Reply:
Correct, from a user perspective they'll see a config page with values 80 and 443 by default (which they should be able to change), and a button to "add another".

Additional values would get generated names, yes.  It's the item name + 6 character short uuid, IE `service_ports-8d7hne`.  These key names can be anything as long as they're unique.

Reply:
Two questions.
1. So the vendor choose the names for the default ones, but the additional ones will be generated?
2. The end user can remove defaults, right? So the default could be two ports, 80 and 443, but the end user could change it to just one port, for example.

Reply:
1.  That is how it works today, yes.
2. The user can remove anything but the last value.  There must always be at least 1 value provided or rendering will break.  There is a 'minimumCount' field to prevent deleting too many values (if the vendor required at least 3 ports, for example), but I'd like to do some more robust testing on that before adding it to the docs.